pints/_mcmc/_slice_stepout.py

```diff
@@ -0,0 +1,681 @@
+    sampling scheme, overrelaxed updates are alternated with normal Stepout
+    updates. To obtain the full benefits of overrelaxation, [1] suggests to
+    set almost every update to being overrelaxed and to set the limit ``m``
+    for finding ``I`` to infinity. The algorithm consists of the following
+    steps:
+
+    1. ``\bar{L} = L``,``\bar{R} = R``,``\bar{w} = w``, ``\bar{a} = a``
+    2. while ``R - L < 1.1 * w``:
+        a. ``M = (\bar{L} + \bar{R}) / 2``
+        b. if ``\bar{a} = 0 `` or ``y < f(M)``, exit loop
+        c. if ``x_0 > M``, ``\bar{L} = M``
+           else, ``\bar{R} = M``
+        d. ``\bar{a} = \bar{a} - 1``
+        e. ``\bar{w} = \bar{w} / 2``
+    3. ``\hat{L} = \bar{L}``, ``\hat{R} = \bar{R}``
+    4. while ``\bar{a} > 0``:
+        a. ``\bar{a} = \bar{a} - 1``
+        b. ``\bar{w} = \bar{w} \ 2``
+        c. if ``y >= f(\hat{L} + \bar{w})``, ``\hat{L} = \hat{L} + \bar{w}``
+        d. if ``y >= f(\hat{R} - \bar{w})``, ``\hat{R} = \hat{R} - \bar{W}``
+    5. ``x_1 = \hat{L} + \hat{R} - x_0``
+    6. if ``x_1 < \bar{L}`` or ``x_1 >= \bar{R}`` or ``y >= f(x_1)``, x_1 = x_0
+
+    The probability of pursuing an overrelaxed step, ``_prob_overrelaxed`` (0
+    as default), and the number of bisection iterations, ``a``, are
+    hyperparameters.
+
+    To avoid floating-point underflow, we implement the suggestion advanced
+    in [1] pp.712. We use the log pdf of the un-normalised posterior
+    (``g(x) = log(f(x))``) instead of ``f(x)``. In doing so, we use an
+    auxiliary variable ``z = log(y) = g(x0) − \epsilon``, where
+    ``\epsilon \sim \text{exp}(1)`` and define the slice as
+    S = {x : z < g(x)}.
+
+    [1] Neal, R.M., 2003. Slice sampling. The annals of statistics, 31(3),
+    pp.705-767.
+    """
+
+    def __init__(self, x0, sigma0=None):
+        super(SliceStepoutMCMC, self).__init__(x0, sigma0)
+
+        # Set initial state
+        self._x0 = np.asarray(x0, dtype=float)
+        self._running = False
+        self._ready_for_tell = False
+        self._current = None
+        self._current_log_pdf = None
+        self._current_log_y = None
+        self._proposed = None
+        self._overrelaxed_step = False
+
+        # Default initial interval width w used in the Stepout procedure
+        # to expand the interval
+        self._w = np.abs(self._x0)
+        self._w[self._w == 0] = 1
+        self._w = 0.1 * self._w
+
+        # Default integer limiting the size of the interval to ``m * w```
+        self._m = 50
+
+        # Flag to initialise the expansion of the interval ``I=(L,R)``
+        self._first_expansion = False
+
+        # Flag indicating whether the interval expansion is concluded
+        self._interval_found = False
+
+        # Number of steps used for expanding the interval ``I``
+        self._j = None
+        self._k = None
+
+        # Flags used to calculate log_pdf of initial interval edges ``l,r```
```

Comment:
There's a lot of values stored here!
Are all of these needed (for example, they get set in ask() and then used in tell())? Or can some of them just be local variables in the methods that use them? Please check for both classes